### PR TITLE
Hw1

### PR DESCRIPTION
**Pull request by Tetiana-KH**

### CODE REVIEW

File: pages/checkout.js
Comment (adudyak) [inline pages/checkout.js:1]: move content of checkout page to cart page, then delete this file

File: pages/cart.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  totalPrice: { xpath: '//*[@id="cart"]/ul/li[2]/div/div[3]/div[2]' },
+  flatShippingRate: { xpath: '//*[@id="collapse-shipping-method"]/div/p[3]' },
+
+  payForProduct() {
+    I.click(this.termsCheckBox);
+    I.click(this.continuePaymentButton);
+  },
+
+  parsePrice(priceString) {
+    return parseFloat(priceString.replace(/[^0-9.-]/g, ''));
```
Comment (adudyak) [inline pages/cart.js]: why do you need `-` in regex? only digits, dots and maybe comma is enough.

File: pages/cart.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  },
+
+  async getTotalPrice() {
+  return this.parsePrice(await I.grabTextFrom(this.totalPrice));
+  },
+
+  async getDelivery() {
```
Comment (adudyak) [inline pages/cart.js]: this method is not used in scenario

File: pages/cart.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    console.log('subTotal:', this.parsePrice(await I.grabTextFrom(this.subTotal)));
+    console.log('totalPrice:', this.parsePrice(await I.grabTextFrom(this.totalPrice)));
+    console.log('flatShippingRate:', this.parsePrice(await I.grabTextFrom(this.flatShippingRate)));
+    return this.parsePrice(await I.grabTextFrom(this.subTotal) + this.parsePrice(await I.grabTextFrom(this.flatShippingRate)));
+  },
+
+  verifySuccessfulPurchase() {
```
Comment (adudyak) [inline pages/cart.js]: this method is not used in scenario

File: pages/success.js
Comment (adudyak) [inline pages/success.js:1]: move content to cart page and delete this file

File: pages/success.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
+const { I } = inject();
+
+module.exports = {
+  successContinueButton: { xpath: '//*[@id="content"]/div/div/a' },
+
+  verifySuccessfulPurchase() {
+    I.amOnPage('/index.php?route=checkout/success');
```
Comment (adudyak) [inline pages/success.js]: you should not navigate to success page, website should navigate you

File: pages/success.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+    const regTitleText = 'Your order has been placed!';
+    I.seeTextEquals(regTitleText, this.h1);
+    I.click(this.successContinueButton);
```
Comment (adudyak) [inline pages/success.js]: remove. method name should correspond to it's instructions. Navigation does not verify success

File: tests/buyProduct_test.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+//Before(({ I }) => {
+  //  I.login(USER);
+//});
+
+Scenario('buy product', async ({ I, cartPage, productPage, checkoutPage, successPage }) => {
+    I.login(USER);
+    I.amOnPage('/index.php?route=product/product&product_id=44');
+    productPage.selectColor();
+    productPage.selectSize();
+    productPage.addToCart();
```
Comment (adudyak) [inline tests/buyProduct_test.js]: `addToCart();` method proceeds to cart page, but next step is grabbing of price, color, size, which are not available in cart page.
Grab prices, then add to cart and proceed to checkout

File: pages/product.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+  parsePrice(priceString) {
+    console.log('parsed price:', parseFloat(priceString.replace(/[^0-9.-]/g, '')));
+    return parseFloat(priceString.replace(/[^0-9.-]/g, ''));
+  },
+
+  async getProductPrice() {
+    return this.parsePrice(await I.grabTextFrom(this.productPriceText)) + this.parsePrice(await I.grabTextFrom(this.colorOption)) + this.parsePrice(await I.grabTextFrom(this.sizeOption));
+  },
+
+  addToCart() {
```
Comment (adudyak) [inline pages/product.js:33]: this method, according to name, should only add to cart, but it is also proceeds to shopping cart. Move extra functionality to other method.

File: pages/product.js [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    I.click(this.addToCartButton);
+    I.click(this.myAccountSpoiler);
+    I.click(this.shoppingCartButton);
```
Comment (adudyak) [inline pages/product.js]: you need to open checkout page, not shopping cart one

File: steps_file.js [hunk 9]
```diff
@@ -1,10 +1,48 @@
...
+      this.amOnPage('/');
+      this.click(signInButton);
+      this.fillField(email, user.email);
+      this.fillField(password, user.password);
+      this.click(loginButton);
+      this.seeTextEquals("Мої замовлення", myOrdersText);
+    },
+
+  openCart() {
+    this.amOnPage('/index.php?route=product/product&product_id=44');
```
Comment (adudyak) [inline steps_file.js]: remove this line. does not correspond to method name and navigation to specific page should be done in separate method or in scenario

File: steps_file.js [hunk 10]
```diff
@@ -1,10 +1,48 @@
...
+    this.click(myAccountSpoiler);
+    this.click(shoppingCartButton);
+  },
 
-  });
+  proceedToCheckOut() {
```
Comment (adudyak) [inline steps_file.js]: move this to cart page.
name it `fillCheckoutForm()`

File: steps_file.js [hunk 11]
```diff
@@ -1,10 +1,48 @@
...
+    this.fillField(city, "Dubai");
+    this.fillField(postCode, "08711");
+    this.click(country);
+    this.click(stateOption);
+    this.click(continueButton);
+    this.click(continueButtonDeliveryDetails);
+    this.click(continueButtonDeliveryMethod);
+    this.click(termsCheckbox);
+    this.click(continuePaymentButton);
+    this.click(confirmOrderButton);
```
Comment (adudyak) [inline steps_file.js]: you need to grab total price, tax before confirming an order

File: tests/buyProduct_test.js [hunk 12]
```diff
@@ -0,0 +1,35 @@
+const { pause } = require("codeceptjs");
+const { flatShippingRate, proceedToCheckOut, subTotal } = require("../pages/cart");
+const { productPriceText } = require("../pages/product");
+const { totalPrice1 } = require("../pages/checkout");
```
Comment (adudyak) [inline tests/buyProduct_test.js]: remove lines 1-4

File: tests/buyProduct_test.js [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+    city: "Dubai",
+    postCode: "08711",
+    address1: "Dubai str 1",
+};
+
+Feature('buy product');
+
+//Before(({ I }) => {
+  //  I.login(USER);
+//});
```
Comment (adudyak) [inline tests/buyProduct_test.js]: uncomment login here, and remove it in scenario

File: tests/buyProduct_test.js [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+Scenario('buy product', async ({ I, cartPage, productPage, checkoutPage, successPage }) => {
+    I.login(USER);
+    I.amOnPage('/index.php?route=product/product&product_id=44');
+    productPage.selectColor();
+    productPage.selectSize();
+    productPage.addToCart();
+    const productPrice = await productPage.getProductPrice();
+    console.log('productPrice: ', productPrice);
+    const totalPrice = await cartPage.getTotalPrice();
+    I.assertEqual(subTotal + flatShippingRate, totalPrice1, "Prices are not in match!");
```
Comment (adudyak) [inline tests/buyProduct_test.js]: `totalPrice` extracted, `totalPrice1` used

Comment (adudyak) [inline tests/buyProduct_test.js]: `flatShippingRate` is not grabbed

File: tests/buyProduct_test.js [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+    I.amOnPage('http://opencart.qatestlab.net/index.php?route=checkout/cart');
```
Comment (adudyak) [inline tests/buyProduct_test.js]: proceed to checkout via buttons in cart.js

File: tests/buyProduct_test.js [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+    address1: "Dubai str 1",
+};
+
+Feature('buy product');
+
+//Before(({ I }) => {
+  //  I.login(USER);
+//});
+
+Scenario('buy product', async ({ I, cartPage, productPage, checkoutPage, successPage }) => {
```
Comment (adudyak) [inline tests/buyProduct_test.js]: update scenario to perform instructions in following order:

1. Open product.
2. Select size, color. Grab productPrice: product + color + size.
3. Add product to cart
4. Open checkout
5. Proceed to last step
6. Grab totalPrice, tax
7. Compare productPrice + tax with totalPrice
8. Complete order
9. Verify successful message is displayed

**Review by adudyak — CHANGES_REQUESTED** [review]

File: pages/success.js [hunk 17]
```diff
@@ -0,0 +1,12 @@
+const { I } = inject();
+
+module.exports = {
+  successContinueButton: { xpath: '//*[@id="content"]/div/div/a' },
+
+  verifySuccessfulPurchase() {
+    I.amOnPage('/index.php?route=checkout/success');
+    const regTitleText = 'Your order has been placed!';
```
Comment (adudyak) [inline pages/success.js:8]: you are using UA text to verify login, but EN to verify purchase.
Add language selection in case site shows language randomly.

**Review by adudyak — CHANGES_REQUESTED** [review]

File: codecept.conf.js [hunk 18]
```diff
@@ -21,12 +21,19 @@ exports.config = {
...
+      require: "codeceptjs-chai"
     }
   },
   include: {
     I: './steps_file.js',
     basePage: "./pages/base.js",
     accountPage: "./pages/account.js",
+    productPage: "./pages/product.js",
+    cartPage: "./pages/cart.js",
+    checkoutPage: "./pages/checkout.js",
```
Comment (adudyak) [inline codecept.conf.js]: remove

File: pages/cart.js [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+    I.waitForVisible('//*[@id="modal-shipping"]/div/div/div[2]/div/label', 5);
+    I.click(this.flatShippingRateRadioButton);
+    I.click(this.applyShippingButton);
+    I.click(this.checkOutButton);
+    I.click(this.deliveryDetailsContinueButton);
+    I.click(this.deliveryMethodContinueButton);
+    I.click(this.termsCheckbox);
+    I.click(this.paymentMethodContinueButton);
+    const totalPrice = await cartPage.getTotalPrice();
+    I.assertEqual(subTotal + flatShippingRate, totalPrice, "Prices are not in match!");
```
Comment (adudyak) [inline pages/cart.js]: remove, assert should be only in scenario

File: pages/cart.js [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+  },
+
+  async getTotalPrice() {
+    return this.parsePrice(await I.grabTextFrom(this.totalPrice));
+  },
+
+  async proceedToCheckOut() {
+    I.click(this.shippingAndTaxesDropDown);
+    I.click(this.getQuotesButton);
+    I.waitForVisible('//*[@id="modal-shipping"]/div/div/div[2]/div/label', 5);
```
Comment (adudyak) [inline pages/cart.js]: move locator to top

File: pages/cart.js [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+    I.click(this.flatShippingRateRadioButton);
+    I.click(this.applyShippingButton);
+    I.click(this.checkOutButton);
+    I.click(this.deliveryDetailsContinueButton);
+    I.click(this.deliveryMethodContinueButton);
+    I.click(this.termsCheckbox);
+    I.click(this.paymentMethodContinueButton);
+    const totalPrice = await cartPage.getTotalPrice();
```
Comment (adudyak) [inline pages/cart.js]: remove

File: pages/cart.js [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+    I.click(this.confirmOrderButton);
+  },
+
+  parsePrice(priceString) {
+    console.log('parsed price:', parseFloat(priceString.replace(/[^0-9.]/g, '')));
+    return parseFloat(priceString.replace(/[^0-9.]/g, ''));
+  },
+
+  async getProductPrice() {
+    return this.parsePrice(await I.grabTextFrom(this.subTotalPrice)) + this.parsePrice(await I.grabTextFrom(this.flatShippingRate));
```
Comment (adudyak) [inline pages/cart.js]: during checkout process:
grab shipping rate and return it to scenario, where add to product price + color + price:
image
then compare your calculations to bottom price:
image

File: tests/buyProduct_test.js [hunk 23]
```diff
@@ -0,0 +1,32 @@
...
+});
+
+Scenario('buy product', async ({ I, cartPage, productPage, successPage }) => {
+  I.amOnPage('/index.php?route=product/product&product_id=44');
+  productPage.selectColor();
+  productPage.selectSize();
+  const productPrice = await productPage.getProductPrice();
+  console.log('productPrice: ', productPrice);
+  const totalPrice = await cartPage.getTotalPrice();
+  I.assertEqual(subTotal + flatShippingRate, totalPrice, "Prices are not in match!");
```
Comment (adudyak) [inline tests/buyProduct_test.js]: I don't see grabbing of flatShippingRate, you are using locator from cart page

File: tests/buyProduct_test.js [hunk 24]
```diff
@@ -0,0 +1,32 @@
+const { subTotal, flatShippingRate } = require("../pages/cart");
```
Comment (adudyak) [inline tests/buyProduct_test.js]: remove

File: tests/buyProduct_test.js [hunk 25]
```diff
@@ -0,0 +1,32 @@
...
+const steps_file = require("../steps_file");
```
Comment (adudyak) [inline tests/buyProduct_test.js]: remove

File: steps_file.js [hunk 26]
```diff
@@ -1,10 +1,26 @@
-// in this file you can append custom step methods to 'I' object
+const { pause } = require("codeceptjs");
```
Comment (adudyak) [inline steps_file.js]: remove

File: steps_file.js [hunk 27]
```diff
@@ -1,10 +1,26 @@
...
+const { billingDetailsDropDown, newAddressRadioButton, firstName, lastName, address1, city, postCode, country, stateOption, confirmOrderButton } = require("./pages/cart");
```
Comment (adudyak) [inline steps_file.js]: remove

File: steps_file.js [hunk 28]
```diff
@@ -1,10 +1,26 @@
...
+const { myAccountSpoiler, shoppingCartButton } = require("./pages/product");
```
Comment (adudyak) [inline steps_file.js]: remove

File: steps_file.js [hunk 29]
```diff
@@ -1,10 +1,26 @@
...
+const { languageSelectDropDown, selectEnglishButton } = require("./pages/account");
```
Comment (adudyak) [inline steps_file.js]: it is strange, that you store locators, which only used here, in other object.
move them here, as the ones below

File: steps_file.js [hunk 30]
```diff
@@ -1,10 +1,26 @@
...
 
-module.exports = function() {
-  return actor({
+email = { css: "#input-email" };
```
Comment (adudyak) [inline steps_file.js]: add `const` to variables, these are not object keys

**Review by adudyak — CHANGES_REQUESTED** [review]

File: tests/buyProduct_test.js [hunk 31]
```diff
@@ -0,0 +1,32 @@
...
+
+Scenario('buy product', async ({ I, cartPage, productPage, successPage }) => {
+  I.amOnPage('/index.php?route=product/product&product_id=44');
+  productPage.selectColor();
+  productPage.selectSize();
+  const productPrice = await productPage.getProductPrice();
+  console.log('productPrice: ', productPrice);
+  const totalPrice = await cartPage.getTotalPrice();
+  I.assertEqual(subTotal + flatShippingRate, totalPrice, "Prices are not in match!");
+  productPage.addToCart();
```
Comment (adudyak) [inline tests/buyProduct_test.js:22]: add product to cart and open cart before extracting totalPrice from cart. Currently it is vice versa

**Review by adudyak — CHANGES_REQUESTED** [review]

File: pages/cart.js (hunk 22 above)
Comment (Tetiana-KH) [reply pages/cart.js]: Not sure if I've done this correctly

File: pages/cart.js [hunk 32]
```diff
@@ -0,0 +1,71 @@
...
+    I.click(this.paymentMethodContinueButton);
+    I.click(this.confirmOrderButton);
+  },
+
+  parsePrice(priceString) {
+    console.log('parsed price:', parseFloat(priceString.replace(/[^0-9.]/g, '')));
+    return parseFloat(priceString.replace(/[^0-9.]/g, ''));
+  },
+
+  async getTotalPrice() {
```
Comment (adudyak) [inline pages/cart.js]: you already have method with same name

File: tests/buyProduct_test.js [hunk 33]
```diff
@@ -0,0 +1,31 @@
...
+  address1: "Dubai str 1",
+};
+
+Feature('buy product');
+
+Before(({ I }) => {
+  I.login(USER);
+});
+
+Scenario('buy product', async ({ I, cartPage, productPage, successPage }) => {
```
Comment (adudyak) [inline tests/buyProduct_test.js]: please check navigation order

**Review by adudyak — CHANGES_REQUESTED** [review]

File: pages/cart.js (hunk 32 above)
Comment (Tetiana-KH) [reply pages/cart.js]: removed this duplicate

File: tests/buyProduct_test.js (hunk 33 above)
Comment (Tetiana-KH) [reply tests/buyProduct_test.js]: corrected in the following order: I, productPage, cartPage, successPage

File: pages/cart.js [hunk 34]
```diff
@@ -0,0 +1,57 @@
...
+  async proceedToCheckOut() {
+    I.click(this.cartIcon);
+    I.click(this.checkOutButton);
+    I.click(this.billingDetailsContinueButton);
+    I.click(this.deliveryDetailsContinueButton);
+    I.waitForVisible(this.flatRateText, 5);
+    I.click(this.deliveryMethodContinueButton);
+    I.click(this.termsCheckbox);
+    I.click(this.paymentMethodContinueButton);
+    I.click(this.confirmOrderButton);
```
Comment (adudyak) [inline pages/cart.js]: move this line to separate method and call after comparing prices

File: pages/product.js [hunk 35]
```diff
@@ -0,0 +1,36 @@
...
+    I.click(this.sizeOption);
+  },
+
+  parsePrice(priceString) {
+    console.log('parsed price:', parseFloat(priceString.replace(/[^0-9.]/g, '')));
+    return parseFloat(priceString.replace(/[^0-9.]/g, ''));
+  },
+
+  async getProductPrice() {
+    return this.parsePrice(await I.grabTextFrom(this.productPriceText)) + this.parsePrice(await I.grabTextFrom(this.colorOption)) + this.parsePrice(await I.grabTextFrom(this.sizeOption)) + this.parsePrice(await I.grabTextFrom(this.flatShippingRate));
```
Comment (adudyak) [inline pages/product.js]: remove everything after size, it is not available at product page

File: tests/buyProduct_test.js [hunk 36]
```diff
@@ -0,0 +1,28 @@
...
+});
+
+Scenario('buy product', async ({ I, productPage, cartPage, successPage }) => {
+  I.amOnPage('/index.php?route=product/product&path=260&product_id=48');
+  productPage.selectColor();
+  productPage.selectSize();
+  const productPrice = await productPage.getProductPrice();
+  productPage.addToCart();
+  cartPage.proceedToCheckOut();
+  const flatShippingRate = await cartPage.getFlatShippingRate();
```
Comment (adudyak) [inline tests/buyProduct_test.js:24]: `flatShippingRate` is equal to `NaN`. Check locator

File: tests/buyProduct_test.js [hunk 37]
```diff
@@ -0,0 +1,28 @@
...
+  const totalPrice = await cartPage.getTotalPrice();
```
Comment (adudyak) [inline tests/buyProduct_test.js:25]: `totalPrice` is `220`, but should be `225` for product 48. Check locator

**Review by adudyak — CHANGES_REQUESTED** [review]

File: pages/cart.js [hunk 38]
```diff
@@ -0,0 +1,53 @@
...
+  cartIcon: { xpath: '//*[@id="cart"]/button' },
+  termsCheckBox: { xpath: '//*[@id="agree1"]' },
+  continuePaymentButton: { xpath: '//*[@id="button-payment-method"]' },
+  subTotal: { xpath: '//strong[text()="Sub-Total:"]/parent::td/following-sibling::td' },
+  totalPrice: { xpath: '//strong[text()="Total:"]/parent::td/following-sibling::td' },
+  flatShippingRate: { xpath: '//strong[text()="Flat Shipping Rate:"]/parent::td/following-sibling::td' },
+  shippingAndTaxesDropDown: { xpath: '//*[@id="accordion"]/div[2]/div[1]/h4/a' },
+  getQuotesButton: { xpath: '//*[@id="button-quote"]' },
+  flatRateText: { xpath: '//*[@id="collapse-shipping-method"]/div/p[2]' },
+  checkOutButton: { xpath: '//ul[@id="cart"]//li/div/a[contains(text(), "Checkout")]' },
```
Comment (adudyak) [inline pages/cart.js:13]: this locator failed for me, try simple version:
`'//div/a[contains(text(), "Checkout")]'`

**Review by adudyak — APPROVED** [review]

File: pages/cart.js (hunk 38 above)
Comment (Tetiana-KH) [reply pages/cart.js:13]: Updated

**Pull request merged**